# Chennai PG Intelligence: Exploratory Data Analysis

## importing modules

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# PHASE 0: Data Loading

In [2]:
df = pd.read_csv(r'D:\Hustle\Chennai-PG\Data\raw\chennai_pg_dataset.csv') # this is raw dataset, DO NOT modify anything

----------------------------------------------------------------------------------
# PHASE 1: SANITY CHECK

- Shape and structure
- Check the datatypes
- Missing Values
- Duplicates
- Unique value count for (categorical)
- Basic stats

In [3]:
df.head()

,id,title,latitude,longitude,locality,address,gender,available_for,transit_score,lifestyle_score,...,room_ac,room_cupboard,room_tv,room_geyser,room_bedding,room_attached_bath,gate_closing_time,smoking_allowed,guardian_required,nonveg_allowed
0,8aa9ba0d9ef7b08b019ef7beaaeb030e,PG for Boys in Anbu Homoeopathy And Acupunctur...,12.892600,80.114600,Perungalathur,"ramasamy nagar , Alapakkam, Anbu Homoeopathy A...",MALE,Both,5.7,4.1,...,True,True,True,True,True,True,NaN,True,True,False
1,8aa9ba0d9ef7b08b019ef7beaaeb030e,PG for Boys in Anbu Homoeopathy And Acupunctur...,12.892600,80.114600,Perungalathur,"ramasamy nagar , Alapakkam, Anbu Homoeopathy A...",MALE,Both,5.7,4.1,...,True,True,True,True,True,True,NaN,True,True,False
2,8aa9ba0d9ef7b08b019ef7beaaeb030e,PG for Boys in Anbu Homoeopathy And Acupunctur...,12.892600,80.114600,Perungalathur,"ramasamy nagar , Alapakkam, Anbu Homoeopathy A...",MALE,Both,5.7,4.1,...,True,True,True,True,True,True,NaN,True,True,False
3,8a9fae838be6a66a018be7e0229327a9,PG for Boys in Perungalathur,12.905105,80.095660,Perungalathur,", Gateway Office Parks Rd near Perungalathur P...",MALE,Both,NaN,NaN,...,False,False,False,False,False,False,NaN,False,False,False
4,8a9fb38499ce0d630199ce1448fd00e1,PG for Boys in Old Perungalathur,12.918673,80.081292,Shanthi Nagar,Parvathi Nagar near ( GANAPATHY MEN'S HOSTEL ...,MALE,Both,5.9,9.1,...,False,False,False,False,False,False,NaN,False,False,False


In [4]:
# 1.1 shape and structure + dtypes
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1783 entries, 0 to 1782
Data columns (total 39 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  1783 non-null   str    
 1   title               1783 non-null   str    
 2   latitude            1783 non-null   float64
 3   longitude           1783 non-null   float64
 4   locality            1783 non-null   str    
 5   address             1783 non-null   str    
 6   gender              1783 non-null   str    
 7   available_for       1783 non-null   str    
 8   transit_score       883 non-null    float64
 9   lifestyle_score     883 non-null    float64
 10  occupancy           1759 non-null   str    
 11  rent                1759 non-null   float64
 12  deposit             1759 non-null   float64
 13  attached_bathroom   1759 non-null   object 
 14  food_included       1783 non-null   bool   
 15  breakfast           1783 non-null   bool   
 16  lunch            

In [5]:
# 1.2 Missing values
def missing_info(df):
    missing_info = pd.DataFrame({
        'missing count': df.isna().sum(),
        'missing %': (df.isna().mean() * 100).round(2),
        'Data type': df.dtypes,
    })

    missing_info = missing_info[missing_info['missing count'] > 0]
    missing_info.sort_values('missing %', ascending = False)
    return missing_info

missing_info(df)

,missing count,missing %,Data type
transit_score,900,50.48,float64
lifestyle_score,900,50.48,float64
occupancy,24,1.35,str
rent,24,1.35,float64
deposit,24,1.35,float64
attached_bathroom,24,1.35,object
mess,834,46.78,object
wifi,830,46.55,object
laundry,806,45.20,object
power_backup,830,46.55,object


In [8]:
# 1.3 Duplicates
def isDuplicates(df: pd.DataFrame):
    print(f'Total duplicates: {df.duplicated().sum()}')
    return df[df.duplicated(keep=False)].sort_values(by=df.columns.tolist()).head(20)

isDuplicates(df)

Total duplicates: 162


,id,title,latitude,longitude,locality,address,gender,available_for,transit_score,lifestyle_score,...,room_ac,room_cupboard,room_tv,room_geyser,room_bedding,room_attached_bath,gate_closing_time,smoking_allowed,guardian_required,nonveg_allowed
830,8a9f82038cb9ca57018cbaea42625878,PG for Boys in Tharamani,12.982579,80.238133,Tharamani,"Anna Nagar near Panchamukha Hanuman Temple, Th...",MALE,Both,NaN,NaN,...,False,True,False,False,True,True,11:00 PM,True,True,False
833,8a9f82038cb9ca57018cbaea42625878,PG for Boys in Tharamani,12.982579,80.238133,Tharamani,"Anna Nagar near Panchamukha Hanuman Temple, Th...",MALE,Both,NaN,NaN,...,False,True,False,False,True,True,11:00 PM,True,True,False
832,8a9f82038cb9ca57018cbaea42625878,PG for Boys in Tharamani,12.982579,80.238133,Tharamani,"Anna Nagar near Panchamukha Hanuman Temple, Th...",MALE,Both,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,11:00 PM,True,True,False
835,8a9f82038cb9ca57018cbaea42625878,PG for Boys in Tharamani,12.982579,80.238133,Tharamani,"Anna Nagar near Panchamukha Hanuman Temple, Th...",MALE,Both,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,11:00 PM,True,True,False
831,8a9f82038cb9ca57018cbaea42625878,PG for Boys in Tharamani,12.982579,80.238133,Tharamani,"Anna Nagar near Panchamukha Hanuman Temple, Th...",MALE,Both,NaN,NaN,...,False,True,False,False,True,True,11:00 PM,True,True,False
834,8a9f82038cb9ca57018cbaea42625878,PG for Boys in Tharamani,12.982579,80.238133,Tharamani,"Anna Nagar near Panchamukha Hanuman Temple, Th...",MALE,Both,NaN,NaN,...,False,True,False,False,True,True,11:00 PM,True,True,False
1013,8a9f84449233260d019233676f89127e,PG for Girls in Tharamani,12.982616,80.238006,Tharamani,"Anna Nagar, near Panchamukha Hanuman Temple, ...",FEMALE,Both,NaN,NaN,...,False,False,False,False,False,False,NaN,False,False,False
1016,8a9f84449233260d019233676f89127e,PG for Girls in Tharamani,12.982616,80.238006,Tharamani,"Anna Nagar, near Panchamukha Hanuman Temple, ...",FEMALE,Both,NaN,NaN,...,False,False,False,False,False,False,NaN,False,False,False
1015,8a9f84449233260d019233676f89127e,PG for Girls in Tharamani,12.982616,80.238006,Tharamani,"Anna Nagar, near Panchamukha Hanuman Temple, ...",FEMALE,Both,NaN,NaN,...,False,False,False,False,False,False,NaN,False,False,False
1018,8a9f84449233260d019233676f89127e,PG for Girls in Tharamani,12.982616,80.238006,Tharamani,"Anna Nagar, near Panchamukha Hanuman Temple, ...",FEMALE,Both,NaN,NaN,...,False,False,False,False,False,False,NaN,False,False,False


In [9]:
# 1.4 Unique valu counts (categorical)
for col in df.select_dtypes(include='object').columns:
    print(f'{col}: {df[col].nunique()} unique -> {df[col].unique()[:10]}')

id: 657 unique -> <StringArray>
['8aa9ba0d9ef7b08b019ef7beaaeb030e', '8a9fae838be6a66a018be7e0229327a9',
 '8a9fb38499ce0d630199ce1448fd00e1', '8aa98e6a9c1c4dca019c1cc70bf621a8',
 '8aa98e979b82e1a4019b83d99c297a01', '8a9f9182844bc06501844c39248d6b15',
 'ff808181547b7f2801547f8dfa492a85', '8a9fa98495832f000195836bfe450c9d',
 '8aa9aa7e9c64d498019c65163c521ee1', '8a9fae85971a0e6d01971a5203a7105b']
Length: 10, dtype: str
title: 181 unique -> <StringArray>
['PG for Boys in Anbu Homoeopathy And Acupuncture Clinic',
                           'PG for Boys in Perungalathur',
                       'PG for Boys in Old Perungalathur',
                         'PG for Boys in Shollinganallur',
                       'PG for Boys in New Perungalathur',
                          'PG for Boys in Perungalathur,',
                          'PG for Boys in West Tambaram ',
                            'PG for Boys in Third Street',
                        'PG for Boys in Otteri Extension',
              

C:\Users\Dinesh\AppData\Local\Temp\ipykernel_16508\2567202304.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for col in df.select_dtypes(include='object').columns:


In [12]:
# 1.5 Basic stats (numerical columns)
df.describe(include=['int64', 'float64'])   # numeric columns

,latitude,longitude,transit_score,lifestyle_score,rent,deposit,total_bathrooms
count,1783.000000,1783.000000,883.000000,883.000000,1759.000000,1759.000000,1783.000000
mean,12.959144,80.208051,7.053001,6.934428,7207.355884,5881.930642,15.365676
std,0.048097,0.050124,2.016839,1.316050,4126.793881,9785.376485,35.519123
min,12.840608,80.040125,-10.000000,1.500000,0.000000,0.000000,0.000000
25%,12.913428,80.212143,6.400000,6.200000,6000.000000,3000.000000,0.000000
50%,12.977430,80.228473,7.600000,6.300000,7000.000000,3000.000000,0.000000
75%,12.985703,80.235708,8.200000,8.100000,8500.000000,5900.000000,12.000000
max,13.115971,80.261954,9.100000,9.600000,65000.000000,200000.000000,215.000000


In [11]:
# 1.5 basic stats (categorical column)
df.describe(include=['object', 'str']) # categorical column

,id,title,locality,address,gender,available_for,occupancy,attached_bathroom,mess,wifi,...,warden,cooking_allowed,parking,room_ac,room_cupboard,room_tv,room_geyser,room_bedding,room_attached_bath,gate_closing_time
count,1783,1783,1783,1783,1783,1783,1759,1759,949,953,...,977,950,1610,1608,1620,1595,1608,1618,1618,307
unique,657,181,36,624,3,3,4,2,2,2,...,2,2,3,2,2,2,2,2,2,24
top,8a9fd32c84898be701848a319d937c89,PG for Boys in Karapakkam,Tharamani,"Tidel park , Tharamani, Tharamani, chennai",MALE,Both,THREE,False,True,True,...,True,False,Bike,False,False,False,False,False,False,10:30 PM
freq,12,131,269,12,890,1577,546,1401,491,821,...,527,488,1406,1149,1072,1305,1199,1104,1085,68


# EDA Phase 1: Initial Dataset Inspection Report

> **Purpose:** Understand the raw dataset's structure, quality, and quirks before any cleaning.
> All observations here drive the decisions executed in `02_Data_Cleaning.ipynb`.
> Feature-level and relationship analysis (Phase 2 onward) runs on the cleaned dataset.

---

## 1.1 Shape, Data Structure & Data Types

### Observations

- The dataset contains missing values across multiple column groups.
- Several boolean columns (wifi, laundry, etc.) are stored as object dtype instead of bool these must be cast before any downstream processing.

### Decision

| Issue | Decision |
|---|---|
| Missing values | Addressed per column group in 1.2 |
| object-typed boolean columns | Cast to bool dtype |

```py
df['wifi'].unique() # to check the obj is string or python True/False as obj
```
---

## 1.2 Missing Values Analysis

### 1.2.1 Score Columns (transit_score, lifestyle_score)

- Both columns are missing **~50%** of values.
- The API simply did not return scores for those listings. The exact reason is unknown the score may be unavailable for that locality or was never computed by NoBroker.
- These are **API-level absences**, not data entry gaps.

**Imputation strategy:**

1. Introduce binary indicator columns to preserve the missingness signal:
   - `transit_score_missing`
   - `lifestyle_score_missing`
2. Impute missing scores using the **median within the same locality** — properties in the same locality share similar accessibility characteristics.
3. If an entire locality has no scores, fill remaining `NaN` values with the **global median**.

---

### 1.2.2 Amenity Columns (`wifi`, `laundry`, and others)

- Amenity columns are missing **~45%** of values.
- Missingness was validated by comparing raw JSON API responses against the live NoBroker listing page for the same properties.

**Validated semantics:**

| Value | Meaning |
|---|---|
| True | PG owner explicitly confirmed the amenity is **available** |
| False | PG owner explicitly confirmed the amenity is **unavailable** |
| NaN | PG owner provided **no information** about this amenity |

**Root cause:** Listings with an empty `amenitiesMap` in the API response provide no amenity information whatsoever. The `NaN` is structural, not accidental.

**Modeling assumption:**
> Missing amenity values are treated as `False` unlisted amenities are assumed unavailable.
> This is a conservative, defensible assumption given the source structure.

---

### 1.2.3 Columns with ~1% Missing Values

- Several columns including the **target variable rent**  have approximately **1% missing values**.
- These rows are **dropped outright**. Losing ~1% of the dataset introduces no meaningful bias and is far safer than imputing the target variable.

---

### 1.2.4 gate_closing_time (~80%+ Missing)

- Over **80%** of values are absent.
- Beyond the missingness, the column carries negligible predictive signal for rent estimation.
- **Decision: drop unconditionally.**

---

### 1.2.5 Decision Summary

| Column Group | Missing % | Decision | Rationale |
|---|---|---|---|
| transit_score, lifestyle_score | ~50% | Retain + locality-median imputation | API-level absence; indicator columns added |
| Amenity columns | ~45% | Impute NaN → False | Validated: missing = owner did not disclose |
| rent + ~1% columns | ~1% | Drop rows | Safe data loss; imputing target is unsound |
| gate_closing_time | ~80%+ | Drop column | High missingness + low predictive value |

---

## 1.3 Duplicate Analysis

### Observations

- Exact duplicate rows were found in the dataset.
- Likely originated from overlapping API responses or listings that appear under multiple localities (e.g., a PG listed under both *Tambaram* and *Perungalathur*).
- Since all columns are identical across duplicates, deduplication is performed on the natural composite key: **id + occupancy**.

### Decision

Duplicates are removed keeping the first occurrence of each unique property-room combination.

```python
df_clean = df.drop_duplicates(subset=['id', 'occupancy'], keep='first')
```

---

## 1.4 Categorical Column Cardinality

| Column | Unique Values | Observation | Decision |
|---|---|---|---|
| id | One per property | Identifier carries no predictive signal | Drop |
| title | High cardinality, free text | No structured information useful for modeling | Drop |
| address | High cardinality, free text | Locality already captures geographic signal | Drop |
| locality | 36 categories | Expected to influence rent; clean and meaningful | Encode |
| gender | 3 categories | Clean, low cardinality | Encode |
| available_for | 3 categories | Clean, low cardinality | Encode |
| occupancy | 4 categories + NaN | 24 NaN rows small enough to drop | Drop NaN rows |
| Amenity columns | True / False / NaN | Treated as per 1.2.2 | Impute + encode |
| parking | Mixed (NaN, bike, car, both) | Re-encode as ordinal or multi-class | Encode as 4 levels: none, bike, car, both |
| gate_closing_time | ~80%+ missing | Already flagged | Drop column |

---

## 1.5 Basic Descriptive Statistics

### Observations

| Column | Issue | Decision |
|---|---|---|
| transit_score | Minimum value of **−10** physically impossible | Drop or impute with NaN, then apply locality-median strategy from 1.2.1 |
| rent | Contains **0** values | Drop target variable cannot be zero, drop rent <= 1000 |
| deposit | Contains **0** values | Investigate; likely waived deposits, handle in cleaning |
| total_bathrooms | Values span 0 → 215 with Q3 = 12 extreme upper outliers | Likely a data entry error; column appears **redundant** → drop |

---

## Phase 1: Preprocessing Contract

The following actions are executed in `02_Data_Cleaning.ipynb`, derived entirely from the observations above.

### Columns to Drop

| Column | Reason |
|---|---|
| id | Identifier, no predictive value |
| title | Free text, no structured signal |
| address | Redundant with locality |
| gate_closing_time | ~80%+ missing + low signal |
| total_bathrooms | Likely erroneous; redundant |

### Rows to Drop

| Condition | Reason |
|---|---|
| rent is NaN or 0 | Target variable must be valid |
| ~1% rows with missing values in key columns | Safe to remove |
| Duplicate (id, occupancy) pairs | Redundant entries from API overlap |
| occupancy is NaN (24 rows) | Too few to impute meaningfully |
### Imputations & Transformations

| Column | Treatment |
|---|---|
| transit_score, lifestyle_score | Add _missing indicator → impute with locality median → fallback to global median |
| Amenity columns | NaN → False |
| Boolean object columns | Cast to bool dtype |
| parking | Re-encode NaN → "none"`, map to 4 levels |

---

> **Next:** `02_Data_Cleaning.ipynb` execute all decisions above and produce the cleaned dataset.
> Feature-level EDA (Phase 2: distributions, correlations, feature vs target) runs on that cleaned output.